### ISSUE TITLE
Allow the crate to be used without the `Send` and `Sync` bounds, behind a feature switch

### ISSUE THREAD

**Issue by jakubadamw** [async-graphql/async-graphql#723]
## Description

It'd be useful for the crate to not impose the `Send` and `Sync` bounds on the user futures and the data they hold on to – which are necessary in the context of multi-threaded runtimes – when the user consciously opts in for a single-threaded runtime.

See #673 for a previous mention of this use case.

This would have to be offered behind a feature switch. Perhaps there could be a `send` feature switch that's on by default or, if a negative switch is preferred, `no_send` that's off by default. That means the bounds would be imposed by default, but if someone wants to opt out, they can.

**Comment by sunli829** [comment]
I hope you can explain why you need this. AFAIK only the single-threaded runtime used by `Actix-web`, but it does not support task stealing.

**Comment by jakubadamw** [comment]
@sunli829, we're using `async_graphql` – which, by the way, is a fantastic crate; thank you! – in a Cloudflare worker using [workers-rs](https://github.com/cloudflare/workers-rs).

Within some of our GraphQL objects' methods we'd like to do work that across an `await` point uses non-`Send` values. These are usually futures originating from 3rd-party crates. With a multithreaded runtime that'd be a problem, but as the one we're targetting is single-threaded, we'd like to be able to lift the requirement.

Rather than wrapping the problematic futures in one that has a false `unsafe impl` for `Send` (relying strictly on the single-threaded nature of the runtime for safety) – which would be an unpleasant hack – we figured the framework itself could or should allow for lifting of this bound when used in runtimes that don't need it.

We'll be happy to contribute a solution to this in code. 🙂

**Comment by sunli829** [comment]
Thank you for explaining this to me, I know that `workers-rs` only has a single-threaded runtime.
Add a feature to make `Async-graphql`  remove the `Send` and `Sync` bounds, this change is very big. 😀

**Comment by NyCodeGHG** [comment]
i also need this feature, i tried implementing it, but i'm not familiar enough with rust macros to do it. Also I did not get to work around this. @jakubadamw could you please show me how you did it if you have it around? That would be great.

**Comment by maraisr** [comment]
Yeah would be cool to support this, Cloudflare Workers is also only a single threaded runtime, so wrapping the env in a Send/Sync with Boxing and all that ceremony is just annoying.

**Comment by omarabid** [comment]
Any progress on this?

**Comment by hamiltop** [comment]
If someone who wanted this could try out my branch and let me know if it is useful that'd be awesome. If it does provide value I will try and make it configurable via feature flags.

Fork with branch: https://github.com/hamiltop/async-graphql/tree/sendless_futures
Draft PR: https://github.com/async-graphql/async-graphql/pull/1372

**Comment by omarabid** [comment]
@hamiltop I started a similar effort here: https://github.com/async-graphql/async-graphql/pull/1122 but yet to find a way to make this feasible through feature-flags without over-complicating the code base. Would be happy to hear your thoughts/ideas on that.

**Comment by hamiltop** [comment]
Nice! Happy to collab on your branch if that's easier. It was actually a good learning exercise so I don't really consider it wasted work. I've got a bit more exploring on feature flags but I'll share some ideas here.
